INDI/INDI/indiserver.c

```diff
@@ -156,16 +155,21 @@ typedef struct
     int pid;            /* process id or REMOTEDVR if remote */
     int rfd;            /* read pipe fd */
     int wfd;            /* write pipe fd */
-    int efd;            /* stderr from driver, if local */
     int restarts;       /* times process has been restarted */
     LilXML *lp;         /* XML parsing context */
     FQ *msgq;           /* Msg queue */
     unsigned int nsent; /* bytes of current Msg sent so far */
-} DvrInfo;
+    struct strDvrInfo* pNextToRestart; /* next to restart, or NULL */
+    int restartDelayus; /* Microseconds before next restart attempt */
+} DvrInfo, *pDvr, **ppDvr;
 static DvrInfo *dvrinfo; /* malloced array of drivers */
 static int ndvrinfo;     /* n total */
 
-static char *me;                                       /* our name */
+#define Mus 1000000     /* Microseconds per second */
+#define SELECT_WAITs 1  /* Select wait, seconds */
+static pDvr pRestarts;  /* linked list of drivers to restart */
```

Comment:
The pointer pRestarts will be NULL when no drivers need to be restarted.

New drivers are added to the end of the restart linked list, and the restartDelayus attributes of all drivers on the list are decremented equally after each select(2) call, so the oldest (smallest) delay will alway be at the front of the list, so the pRestarts pointer value can be safely reassigned the value pRestart->pNextToRestart as the first DvrInfo on the list triggers a restart when its restartDelayus attribute value drops to 0 or less (ten seconds expire since the previous restart attempt for the driver).